README
NAME
    MojoX::Renderer::TT - Template Toolkit renderer for Mojo

SYNOPSIS
    Add the handler:

        sub startup {
            ...

            # Via mojolicious plugin
            $self->plugin(tt_renderer => {template_options => {FILTERS => [ ... ]}});

            # Or manually
            use MojoX::Renderer::TT;

            my $tt = MojoX::Renderer::TT->build(
                mojo => $self,
                template_options => {
                    PROCESS  => 'tpl/wrapper',
                    FILTERS  => [ ... ],
                    UNICODE  => 1,
                    ENCODING => 'UTF-8',
                }
            );

            $self->renderer->add_handler( tt => $tt );
        }

    Template parameter are taken from " $c-"stash >.

RENDERING
    The template file for "example/welcome" would be
    "templates/welcome.html.tt".

    When template file is not available rendering from "__DATA__" is
    attempted.

        __DATA__

        @@ welcome.html.tt
        Welcome, [% user.name %]!

    Inline template is also supported:

        $self->render(inline => '[% 1 + 1 %]', handler => 'tt');

HELPERS
    Helpers are exported automatically under "h" namespace.

        [% h.url_for('index') %]

METHODS
  build
    This method returns a handler for the Mojolicious renderer.

    Supported parameters are

    mojo "build" currently uses a "mojo" parameter pointing to the base
    class (Mojo). object. When used the INCLUDE_PATH will be set to
    template_options
        A hash reference of options that are passed to Template->new().

AUTHOR
    Ask Bjørn Hansen, "<ask at develooper.com>"

TODO
       * Better support non-Mojolicious frameworks
       * Move the default template cache directory?
       * Better way to pass parameters to the templates? (stash)
       * More sophisticated default search path?

BUGS
    Please report any bugs or feature requests to "bug-mojox-renderer-tt at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=MojoX-Renderer-TT>. I
    will be notified, and then you'll automatically be notified of progress
    on your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc MojoX::Renderer::TT

    You can also look for information at:

    *   git repository

        <http://git.develooper.com/?p=MojoX-Renderer-TT.git;a=summary>,
        <git://git.develooper.com/MojoX-Renderer-TT.git>

        <http://github.com/abh/mojox-renderer-tt/>

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=MojoX-Renderer-TT>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/MojoX-Renderer-TT>

    *   Search CPAN

        <http://search.cpan.org/dist/MojoX-Renderer-TT/>

ACKNOWLEDGEMENTS
COPYRIGHT & LICENSE
    Copyright 2008-2010 Ask Bjørn Hansen, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for MojoX-Renderer-TT

1.13    June 7, 2011
        - Compatibility with Mojolicious 1.3+

1.12    February 28, 2011
        - Inline wrappers and includes now work on Windows
          (GH#13, Christiaan Kras)

1.11    February 6, 2011
        - Change to use relative paths (Marcus Ramberg)
        - Add exception template for tests

1.10    February 5, 2011
        - Support Mojolicious 'layout'/'extends' (Maxim Vuets, Marcus
          Ramberg)
        - Fix Strawberry Perl tests (RT#65282, Christiaan Kras, Ask)
        - Update bugtracker URL in META.yml

1.0     January 22, 2011
        - Require (and be compatible with) Mojolicious 1.0+
        - Added inline template support (Viacheslav Tykhanovskyi & Ask)
        - Fix inline rendering and test failures (Marcus Ramberg)
        - Fixed helper compatibility with Mojolicious 1.0 (Árpád Szász)
        - Support new mojolicious error template (Marcus)
        - Fix unicode test (Marcus & Spleenjack)
        - Replaced deprecated Mojo dependence with Mojolicious (spleenjack)

0.40    August 2, 2010
        - Updated for newer Mojo (Viacheslav Tykhanovskyi)
        - Added helpers (Viacheslav Tykhanovskyi)
        - Documentation updates

0.31    September 11, 2009
        - Remove deprecated 'new' method
        - Update MANIFEST (oops, thanks Sebastian)

0.30    September 9, 2009
        - Revert the change of template_path stash key
        - Reflect recent changes in mojo project (v0.991251) (Sebastian Knapp)
        - One test with Mojolicious added (Sebastian Knapp)

0.21    July 29, 2009
        - Update stash key from 'template_path' to 'template' to
          match mojo change (Sebastian Knapp)

0.20    December 9, 2008
        - Updates to work with Mojo 0.9
        - Add default INCLUDE_PATH
        - Made default compile dir not Unix specific (Viacheslav
          Tykhanovskyi)

0.10    November 12, 2008
        - Update to work with named parameters in new version of Mojo
        - Instead of "tx" pass the context in the "c" variable
        - Rename "new" to "build"
        - Add real tests (Viacheslav Tykhanovskyi)
        - Update to work with new error handling in Mojo
          (Viacheslav Tykhanovskyi)

0.02    November 10, 2008
        - Add template_options parameter
        - Fix Template dependency

0.01    November 9, 2008
        - First version, released on an unsuspecting world.



lib/MojoX/Renderer/TT.pm
package MojoX::Renderer::TT;
BEGIN {
  $MojoX::Renderer::TT::VERSION = '1.13';
}

use warnings;
use strict;

use base 'Mojo::Base';

use File::Spec ();
use Mojo::ByteStream 'b';
use Template ();
use Cwd qw/abs_path/;

__PACKAGE__->attr('tt');

sub build {
    my $self = shift->SUPER::new(@_);
    $self->_init(@_);
    return sub { $self->_render(@_) }
}

sub _init {
    my $self = shift;
    my %args = @_;

    #$Template::Parser::DEBUG = 1;

    my $app = delete $args{mojo} || delete $args{app};

    my $dir = $app && $app->home->rel_dir('tmp/ctpl');

    # TODO
    #   take and process options :-)

    my %config = (
        ($app ? (INCLUDE_PATH => abs_path($app->home->rel_dir('templates'))) : ()),
        COMPILE_EXT => '.ttc',
        COMPILE_DIR => ($dir || abs_path(File::Spec->tmpdir)),
        UNICODE     => 1,
        ENCODING    => 'utf-8',
        CACHE_SIZE  => 128,
        RELATIVE    => 1,
        %{$args{template_options} || {}},
    );

    $config{LOAD_TEMPLATES} =
      [MojoX::Renderer::TT::Provider->new(%config, renderer => $app->renderer)]
      unless $config{LOAD_TEMPLATES};

    $self->tt(Template->new(\%config))
      or Carp::croak "Could not initialize Template object: $Template::ERROR";

    return $self;
}

sub _render {
    my ($self, $renderer, $c, $output, $options) = @_;

    # Inline
    my $inline = $options->{inline};

    # Template
    my $t = $renderer->template_name($options);
    $t = 'inline' if defined $inline;
    return unless $t;


    my $helper = MojoX::Renderer::TT::Helper->new(ctx => $c);

    # Purge previous result
    $$output = '';

    my @params = ({%{$c->stash}, c => $c, h => $helper}, $output, {binmode => ':utf8'});
    $self->tt->{SERVICE}->{CONTEXT}->{LOAD_TEMPLATES}->[0]->ctx($c);

    my $ok = $self->tt->process(defined $inline ? \$inline : $t, @params);

    # Error
    unless ($ok) {

        my $e = Mojo::Exception->new(
            $self->tt->error.'',
            $self->tt->service->process(defined $inline ? \$inline : $t));
        $$output = '';
        $c->app->log->error(qq/Template error in "$t": $e/);
        $c->render_exception($e);
        $self->tt->error('');
        return 0;
    }

    return 1;
}

1;    # End of MojoX::Renderer::TT

package
  MojoX::Renderer::TT::Helper;

use strict;
use warnings;

use base 'Mojo::Base';

our $AUTOLOAD;

__PACKAGE__->attr('ctx');

sub AUTOLOAD {
    my $self = shift;

    my $method = $AUTOLOAD;

    return if $method =~ /^[A-Z]+?$/;
    return if $method =~ /^_/;
    return if $method =~ /(?:\:*?)DESTROY$/;

    $method = (split '::' => $method)[-1];

    die qq/Unknown helper: $method/ unless $self->ctx->app->renderer->helpers->{$method};

    return $self->ctx->$method(@_);
}

1;

package
  MojoX::Renderer::TT::Provider;

use strict;
use warnings;

use base 'Template::Provider';

sub new {
    my $class = shift;
    my %params = @_;

    my $renderer = delete $params{renderer};

    my $self = $class->SUPER::new(%params);
    $self->renderer($renderer);
    return $self;
}

sub renderer      { @_ > 1 ? $_[0]->{renderer}      = $_[1] : $_[0]->{renderer} }
sub ctx           { @_ > 1 ? $_[0]->{ctx}           = $_[1] : $_[0]->{ctx} }

sub _template_modified {1}

sub _template_content {
    my $self = shift;
    my ($path) = @_;

    # Convert backslashes to forward slashes to make inline templates work on Windows
    $path =~ s/\\/\//g;
    my ($t) = ($path =~ m{templates\/(.*)$});

    if (-r $path) {
        return $self->SUPER::_template_content(@_);
    }

    # Try DATA section
    elsif (my $d = $self->renderer->get_data_template($self->ctx, $t)) {
        return wantarray ? ($d, '', time) : $d;
    }

    my $data = '';
    my $error = "$path: not found";
    my $mod_date = time;
    return wantarray ? ($data, $error, $mod_date) : $data;
}

1;

__END__

=encoding utf-8

=head1 NAME

MojoX::Renderer::TT - Template Toolkit renderer for Mojo

=head1 SYNOPSIS

Add the handler:

    sub startup {
        ...

        # Via mojolicious plugin
        $self->plugin(tt_renderer => {template_options => {FILTERS => [ ... ]}});

        # Or manually
        use MojoX::Renderer::TT;

        my $tt = MojoX::Renderer::TT->build(
            mojo => $self,
            template_options => {
                PROCESS  => 'tpl/wrapper',
                FILTERS  => [ ... ],
                UNICODE  => 1,
                ENCODING => 'UTF-8',
            }
        );

        $self->renderer->add_handler( tt => $tt );
    }

Template parameter are taken from C< $c->stash >.

=head1 RENDERING

The template file for C<"example/welcome"> would be C<"templates/welcome.html.tt">.

When template file is not available rendering from C<__DATA__> is attempted.

    __DATA__

    @@ welcome.html.tt
    Welcome, [% user.name %]!

Inline template is also supported:

    $self->render(inline => '[% 1 + 1 %]', handler => 'tt');

=head1 HELPERS

Helpers are exported automatically under C<h> namespace.

    [% h.url_for('index') %]

=head1 METHODS

=head2 build

This method returns a handler for the Mojolicious renderer.

Supported parameters are

=over 4

=item mojo
C<build> currently uses a C<mojo> parameter pointing to the base class (Mojo).
object. When used the INCLUDE_PATH will be set to 

=item template_options

A hash reference of options that are passed to Template->new().

=back

=head1 AUTHOR

Ask Bjørn Hansen, C<< <ask at develooper.com> >>

=head1 TODO

   * Better support non-Mojolicious frameworks
   * Move the default template cache directory?
   * Better way to pass parameters to the templates? (stash)
   * More sophisticated default search path?

=head1 BUGS

Please report any bugs or feature requests to C<bug-mojox-renderer-tt at rt.cpan.org>,
or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=MojoX-Renderer-TT>.  I will be
notified, and then you'll automatically be notified of progress on your bug as I
make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc MojoX::Renderer::TT

You can also look for information at:

=over 4

=item * git repository

L<http://git.develooper.com/?p=MojoX-Renderer-TT.git;a=summary>,
L<git://git.develooper.com/MojoX-Renderer-TT.git>

L<http://github.com/abh/mojox-renderer-tt/>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=MojoX-Renderer-TT>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/MojoX-Renderer-TT>

=item * Search CPAN

L<http://search.cpan.org/dist/MojoX-Renderer-TT/>

=back

=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2008-2010 Ask Bjørn Hansen, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'Template Toolkit renderer for Mojo',
  'AUTHOR' => 'Ask Bjørn Hansen <ask@develooper.com>',
  'BUILD_REQUIRES' => {
    'Test::More' => '0'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.31'
  },
  'DISTNAME' => 'MojoX-Renderer-TT',
  'EXE_FILES' => [],
  'LICENSE' => 'perl',
  'NAME' => 'MojoX::Renderer::TT',
  'PREREQ_PM' => {
    'Mojolicious' => '1.3',
    'Template' => '2.18'
  },
  'VERSION' => '1.13',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



